In [1]:
!pip install gdown
!pip install ipython-autotime

In [2]:
%load_ext autotime

time: 134 µs (started: 2021-02-08 08:24:04 +00:00)


In [3]:
!gdown "https://drive.google.com/uc?id=1Qbi954Bwx-PplM8F_7TrB_blcqcB-bF2"

Downloading...
From: https://drive.google.com/uc?id=1Qbi954Bwx-PplM8F_7TrB_blcqcB-bF2
To: /content/train.csv
100% 305k/305k [00:00<00:00, 9.74MB/s]
time: 1.72 s (started: 2021-02-08 08:24:04 +00:00)


In [4]:
!gdown "https://drive.google.com/uc?id=1hWycEy8rQ8e_krGyUhqGQiMBxZtD9SWy"

Downloading...
From: https://drive.google.com/uc?id=1hWycEy8rQ8e_krGyUhqGQiMBxZtD9SWy
To: /content/test.csv
2.47MB [00:00, 72.0MB/s]
time: 1.92 s (started: 2021-02-08 08:24:05 +00:00)


In [5]:
!gdown "https://drive.google.com/uc?id=1j7h-SUy2gMJBds6eQalw4CFZC3O7Pr05"

Downloading...
From: https://drive.google.com/uc?id=1j7h-SUy2gMJBds6eQalw4CFZC3O7Pr05
To: /content/collaboration_network.edgelist
38.8MB [00:00, 82.9MB/s]
time: 3.03 s (started: 2021-02-08 08:24:07 +00:00)


In [6]:
!gdown "https://drive.google.com/uc?id=1AEn8P631y-pq6szgAbTm538oM6waf-5E"

Downloading...
From: https://drive.google.com/uc?id=1AEn8P631y-pq6szgAbTm538oM6waf-5E
To: /content/author_embedding.csv
683MB [00:07, 96.9MB/s]
time: 11.3 s (started: 2021-02-08 08:24:10 +00:00)


In [7]:
!gdown "https://drive.google.com/uc?id=1TD3CkcmbvhIdXkCAWaBpY416qDen2vfD"

Downloading...
From: https://drive.google.com/uc?id=1TD3CkcmbvhIdXkCAWaBpY416qDen2vfD
To: /content/author_papers.txt
23.5MB [00:00, 74.2MB/s]
time: 2.63 s (started: 2021-02-08 08:24:22 +00:00)


In [26]:
import pandas as pd
import numpy as np
import networkx as nx
from sklearn.linear_model import Lasso
from sklearn.model_selection import train_test_split
import xgboost as xgb
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import MinMaxScaler
from pathlib import Path

# read training data
df_train = pd.read_csv('train.csv', dtype={'authorID': np.int64, 'h_index': np.float32})
n_train = df_train.shape[0]

# read test data
df_test = pd.read_csv('test.csv', dtype={'authorID': np.int64})
n_test = df_test.shape[0]

# load the graph
G = nx.read_edgelist('collaboration_network.edgelist',
                     delimiter=' ', nodetype=int)
n_nodes = G.number_of_nodes()
n_edges = G.number_of_edges()
print('Number of nodes:', n_nodes)
print('Number of edges:', n_edges)

# read embeddings of abstracts
embeddings = pd.read_csv("author_embedding.csv", header=None)
embeddings = embeddings.rename(columns={0: "authorID"})

Number of nodes: 231239
Number of edges: 1777338
time: 25.1 s (started: 2021-02-08 08:33:22 +00:00)


In [9]:
# create the training matrix. each author is represented by the average of
# the embeddings of the abstracts of his/her top-cited papers
df_train = df_train.merge(embeddings, on="authorID")

# create the test matrix. each author is represented by the average of
# the embeddings of the abstracts of his/her top-cited papers
df_test = df_test.merge(embeddings, on="authorID")
# X_test = df_test.iloc[:, 2:]

time: 1.55 s (started: 2021-02-08 08:24:49 +00:00)


In [10]:
df_train.head()

authorID  h_index         1  ...       254       255       256
0  2124557175      1.0  0.011884  ... -1.399617  1.284111 -0.491038
1  1964267543      4.0 -0.941447  ... -0.520198  0.506366 -1.664988
2  2762410113      1.0 -1.530304  ... -0.178558 -0.286176 -0.743096
3  2225319093      1.0 -0.129008  ... -0.973226 -0.192451  1.583816
4  2122039117     43.0 -0.255069  ... -0.629376 -1.457717  0.290012

[5 rows x 258 columns]

time: 128 ms (started: 2021-02-08 08:24:50 +00:00)


In [11]:
import ast
import json 

f = open("author_papers.txt", "r")
print("Read author_papers.txt")

# loads the inverted abstracts and stores them as id-abstracts in a dictionary dic and in a folder fw
dic = {}
for l in f:
    if(l == "\n"):
        continue
    author, paper = l.split(":")
    paper = ast.literal_eval(paper)
    dic[author] = paper
with open('author.json', 'w') as outfile:
    json.dump(dic, outfile)

Read author_papers.txt
time: 7.28 s (started: 2021-02-08 08:24:50 +00:00)


In [12]:
df_author_paper = pd.read_json("author.json", orient="index", dtype=str)
df_author_paper.tail()

0             1             2  ...             7    8    9
2908387141  2540479521           nan           nan  ...           nan  nan  nan
2908425732  2553344037           nan           nan  ...           nan  nan  nan
2908436250  2907086791           nan           nan  ...           nan  nan  nan
2908499439  2081432213  2070621672.0  2079679191.0  ...  1641369840.0  nan  nan
2908506980  2148994087           nan           nan  ...           nan  nan  nan

[5 rows x 10 columns]

time: 3.42 s (started: 2021-02-08 08:24:58 +00:00)


In [13]:
from tqdm.notebook import tqdm
tqdm.pandas()

df_author_paper["paper_per_author"] = df_author_paper.progress_apply(lambda x: np.sum(x != "nan"), axis=1)
df_author_paper.head()

0             1  ...             9 paper_per_author
1036332  1510273386  1827736641.0  ...  2209934846.0               10
1101850   133459021   179719743.0  ...  2016084258.0               10
1336878  2122092249  2132109814.0  ...   657471641.0               10
1515524  2141827797  2127085795.0  ...  1981200361.0               10
1606427  1907724546           nan  ...           nan                1

[5 rows x 11 columns]

time: 1min 2s (started: 2021-02-08 08:25:01 +00:00)


In [14]:
df_author_paper_slim = df_author_paper[["paper_per_author"]]
df_author_paper_slim.reset_index(inplace=True)
df_author_paper_slim.rename(columns={"index": "authorID"}, inplace=True)
df_author_paper_slim

/usr/local/lib/python3.6/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


authorID  paper_per_author
0          1036332                10
1          1101850                10
2          1336878                10
3          1515524                10
4          1606427                 1
...            ...               ...
231234  2908387141                 1
231235  2908425732                 1
231236  2908436250                 1
231237  2908499439                 8
231238  2908506980                 1

[231239 rows x 2 columns]

time: 30.5 ms (started: 2021-02-08 08:26:04 +00:00)


In [15]:
author_num_papers_train = df_author_paper_slim[df_author_paper_slim["authorID"].isin(df_train["authorID"])]
author_num_papers_train

authorID  paper_per_author
13         7248981                10
14         7318656                10
19         8845325                 4
20         9723467                10
28        12499814                10
...            ...               ...
231212  2907983075                10
231215  2908006952                 1
231219  2908065712                 1
231222  2908154030                 2
231228  2908220509                 1

[23124 rows x 2 columns]

time: 40 ms (started: 2021-02-08 08:26:04 +00:00)


In [16]:
author_num_papers_test = df_author_paper_slim[df_author_paper_slim["authorID"].isin(df_test["authorID"])]
author_num_papers_test

authorID  paper_per_author
0          1036332                10
1          1101850                10
2          1336878                10
3          1515524                10
4          1606427                 1
...            ...               ...
231234  2908387141                 1
231235  2908425732                 1
231236  2908436250                 1
231237  2908499439                 8
231238  2908506980                 1

[208115 rows x 2 columns]

time: 45.6 ms (started: 2021-02-08 08:26:04 +00:00)


In [17]:
# computes structural features for each node
core_number = nx.core_number(G)
onion_number = nx.onion_layers(G)
avg_neighbor_degree = nx.average_neighbor_degree(G)
degree_centrality = nx.degree_centrality(G)
clustering = nx.clustering(G)
print("Features computed")

Features computed
time: 4min 44s (started: 2021-02-08 08:26:04 +00:00)


In [79]:
# create the training matrix. each node is represented as a vector of 3 features:
# (1) its degree, (2) its core number and (3) the average degree of its neighbors
# X_train_graph = np.zeros((n_train, 12))
X_train_graph = np.zeros((n_train, 8))
y_train_graph = np.zeros(n_train)
for i, row in df_train.iterrows():
    node = row['authorID']
    X_train_graph[i, 0] = G.degree(node)
    X_train_graph[i, 1] = core_number[node]
    X_train_graph[i, 2] = avg_neighbor_degree[node]
    X_train_graph[i, 3] = onion_number[node]
    X_train_graph[i, 4] = degree_centrality[node]
    X_train_graph[i, 5] = clustering[node]
    X_train_graph[i, 6] = row['authorID']
    #for author in G.neighbors(node):
    k_neigbors = df_author_paper_slim[df_author_paper_slim['authorID'].isin(list(G.neighbors(node)))]
    X_train_graph[i, 7] = k_neigbors['paper_per_author'].mean()
    # X_train_graph[i, 8] = k_neigbors['paper_per_author'].std()
    # X_train_graph[i, 9] = k_neigbors['paper_per_author'].median()
    # X_train_graph[i, 10] = k_neigbors['paper_per_author'].quantile(0.25)
    # X_train_graph[i, 11] = k_neigbors['paper_per_author'].quantile(0.75)
    y_train_graph[i] = row['h_index']

time: 2min (started: 2021-02-08 09:49:19 +00:00)


In [ ]:
# X_test_graph = np.zeros((n_test, 12))
X_test_graph = np.zeros((n_test, 8))
for i, row in df_test.iterrows():
    node = row['authorID']
    X_test_graph[i, 0] = G.degree(node)
    X_test_graph[i, 1] = core_number[node]
    X_test_graph[i, 2] = avg_neighbor_degree[node]
    X_test_graph[i, 3] = onion_number[node]
    X_test_graph[i, 4] = degree_centrality[node]
    X_test_graph[i, 5] = clustering[node]
    X_test_graph[i, 6] = row['authorID']
    k_neigbors = df_author_paper_slim[df_author_paper_slim['authorID'].isin(list(G.neighbors(node)))]
    X_test_graph[i, 7] = k_neigbors['paper_per_author'].mean()
    # X_test_graph[i, 8] = k_neigbors['paper_per_author'].std()
    # X_test_graph[i, 9] = k_neigbors['paper_per_author'].median()
    # X_test_graph[i, 10] = k_neigbors['paper_per_author'].quantile(0.25)
    # X_test_graph[i, 11] = k_neigbors['paper_per_author'].quantile(0.75)

In [ ]:
print(X_train_graph.shape)
print(y_train_graph.shape)
print(X_test_graph.shape)

In [ ]:
## Merge text features and graph
# columns=["degree", "core_number", "avg_neighbors", "onion_number", "degree_centrality", "clustering", "authorID",'K_mean','K_std','K_median','K_q1','K_q3']
columns=["degree", "core_number", "avg_neighbors", "onion_number", "degree_centrality", "clustering", "authorID",'K_mean']
X_train_graph_df = pd.DataFrame(
    X_train_graph, 
    columns=columns
)
X_test_graph_df = pd.DataFrame(
    X_test_graph,
    columns=columns
)

In [ ]:
X_train = X_train_graph_df.merge(df_train, on="authorID")
X_test = X_test_graph_df.merge(df_test, on="authorID")

In [ ]:
X_train = X_train.merge(author_num_papers_train, on="authorID")
X_test = X_test.merge(author_num_papers_test, on="authorID")

In [ ]:
X_test

In [ ]:
# Removing h_index and author_id
y_train = X_train["h_index"]
X_train.drop(columns=["authorID", "h_index"], inplace=True)
print(X_train.head())

X_test.drop(columns=["authorID", "h_index_pred"], inplace=True)
print(X_test.head())

# A la francaise 

In [ ]:
# scale for 0 mean and 1 std per col
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

X_train_final, X_test_final, y_train_final, y_test_final = train_test_split(
    X_train_scaled, y_train, test_size=0.2, random_state=42
)

In [ ]:
# train a regression model and make predictions
model = xgb.XGBRegressor()
model.fit(X_train_final, y_train_final)

In [ ]:
y_pred = model.predict(X_test_final)
print(f"Loss: {mean_absolute_error(y_test_final, y_pred)}")

In [ ]:
# Save to file
y_pred_save = model.predict(X_test_scaled)
df_test['h_index_pred'].update(pd.Series(np.round_(y_pred_save, decimals=3)))
df_test.loc[:, ["authorID", "h_index_pred"]].to_csv(
    'predictions.csv', index=False
)

In [ ]:
df_test.loc[:, ["authorID", "h_index_pred"]]

# finetuning 

---



In [ ]:
## Hyper parameter search
train_dmatrix = xgb.DMatrix(data=X_train_final, label=y_train_final)

time: 39.3 ms (started: 2021-02-07 11:03:47 +00:00)


## Early stopping for num of boosting rounds

In [ ]:
# Creata the parameter dictionary for each tree: params
params = {
    "objective":"reg:squarederror", 
    "max_depth": 5,
    "n_estimators": 200
}

# Perform cross-validation with early-stopping: cv_results
cv_results = xgb.cv(
    dtrain=train_dmatrix, 
    nfold=4, 
    params=params, 
    metrics="mae", 
    early_stopping_rounds=10, 
    num_boost_round=100,
    seed=123
)

cv_results

train-mae-mean  train-mae-std  test-mae-mean  test-mae-std
0         6.512381       0.033530       6.530529      0.122788
1         5.230396       0.030879       5.292411      0.131111
2         4.681756       0.027850       4.796288      0.139419
3         4.440768       0.027814       4.615366      0.142417
4         4.314823       0.029789       4.557977      0.140311
5         4.239606       0.027406       4.545356      0.127317
6         4.182877       0.026721       4.539812      0.114327
7         4.144245       0.024540       4.537821      0.108271
8         4.104380       0.019467       4.541241      0.105299
9         4.059746       0.018686       4.538721      0.103680
10        4.011544       0.018462       4.540724      0.098882
11        3.969213       0.015545       4.539878      0.093624
12        3.925572       0.016473       4.536955      0.095537
13        3.888956       0.016689       4.543103      0.088194
14        3.857279       0.022323       4.539264      0.081766
15        3.815661       0.019085       4.526284      0.079085
16        3.781229       0.020461       4.531713      0.083478
17        3.747499       0.024232       4.531349      0.082407
18        3.717136       0.020853       4.520039      0.082373
19        3.684262       0.022220       4.513883      0.077170
20        3.651714       0.026761       4.508451      0.074796
21        3.622872       0.027443       4.499397      0.077157

time: 55.8 s (started: 2021-02-07 11:07:13 +00:00)


## Tunig eta (learning rate)

In [ ]:
# Create list of eta values and empty list to store final round rmse per xgboost model
eta_vals = [0.001, 0.01, 0.1]
best_rmse = []

# Systematicallyvary the eta
for curr_val in eta_vals:
    params['eta'] = curr_val
    
    # Perform cross-validation: cv_results
    cv_results = xgb.cv(dtrain=train_dmatrix, params=params, nfold=4,
                        early_stopping_rounds=5, num_boost_round=10, metrics='mae', seed=123, 
                       as_pandas=True)
    
    # Append the final round rmse to best_rmse
    best_rmse.append(cv_results['test-mae-mean'].tail().values[-1])
    
# Print the result DataFrame
print(pd.DataFrame(list(zip(eta_vals, best_rmse)), columns=['eta', 'best_mae']))

     eta  best_mae
0  0.001  8.974660
1  0.010  8.204939
2  0.100  4.830492
time: 36.8 s (started: 2021-02-06 20:14:08 +00:00)


## Tunig tree depth

In [ ]:
# Create the parameter dictionary
params = {"objective":"reg:squarederror"}

# Create list of max_depth values
max_depths = [2, 5, 10, 20, 50, 100, 500, 1000]
best_rmse = []

for curr_val in max_depths:
    params['max_depth'] = curr_val
    
    # Perform cross-validation
    cv_results = xgb.cv(dtrain=train_dmatrix, params=params, nfold=4, 
                       early_stopping_rounds=10, num_boost_round=50, metrics='mae', seed=123,
                        as_pandas=True)
    
    # Append the final round rmse to best_rmse
    best_rmse.append(cv_results['test-mae-mean'].tail().values[-1])
    
# Print the result DataFrame
print(pd.DataFrame(list(zip(max_depths, best_rmse)), columns=['max_depth', 'best_mae']))

   max_depth  best_mae
0          2  5.020174
1          5  4.499397
2         10  4.505210
3         20  4.602783
4         50  4.619110
5        100  4.619110
6        500  4.619110
7       1000  4.619110
time: 12min 55s (started: 2021-02-06 20:14:45 +00:00)


## Tuning colsample_bytree

In [ ]:
# Create the parameter dictionary
params={"objective":"reg:squarederror", "max_depth":3}

# Create list of hyperparameter values: colsample_bytree_vals
colsample_bytree_vals = [0.1, 0.5, 0.8, 1]
best_rmse = []

# Systematically vary the hyperparameter value 
for curr_val in colsample_bytree_vals:
    params['colsample_bytree'] = curr_val
    
    # Perform cross-validation
    cv_results = xgb.cv(dtrain=train_dmatrix, params=params, nfold=4,
                 num_boost_round=10, early_stopping_rounds=5,
                 metrics="mae", as_pandas=True, seed=123)
    
    # Append the final round rmse to best_rmse
    best_rmse.append(cv_results["test-mae-mean"].tail().values[-1])

# Print the resultant DataFrame
print(pd.DataFrame(list(zip(colsample_bytree_vals, best_rmse)), 
                   columns=["colsample_bytree","best_mae"]))

   colsample_bytree  best_mae
0               0.1  5.618477
1               0.5  4.950673
2               0.8  4.782896
3               1.0  4.796019
time: 29.4 s (started: 2021-02-06 20:27:41 +00:00)


In [ ]:
from sklearn.model_selection import GridSearchCV

# Create the parameter grid: gbm_param_grid
gbm_param_grid = {
    'colsample_bytree': [0.8],
    'n_estimators': [200],
    'max_depth': [5],
    'eta': [0.1, 0.2, 0.5],
    'early_stopping_rounds': [10], 
    'num_boost_round': [10]
}

# Instantiate the regressor: gbm
gbm = xgb.XGBRegressor()

# Perform grid search: grid_mse
grid_mse = GridSearchCV(param_grid=gbm_param_grid, estimator=gbm, 
                        scoring='neg_mean_absolute_error', cv=2, verbose=1, n_jobs=-1)

# Fit grid_mse to the data
grid_mse.fit(X_train_scaled, y_train)

# Print the best parameters and lowest RMSE
print("Best parameters found: ", grid_mse.best_params_)
print("Lowest MAE found: ", np.abs(grid_mse.best_score_))

Fitting 2 folds for each of 3 candidates, totalling 6 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done   6 out of   6 | elapsed:  4.4min finished


[20:45:43] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
Best parameters found:  {'colsample_bytree': 0.8, 'early_stopping_rounds': 10, 'eta': 0.1, 'max_depth': 5, 'n_estimators': 200, 'num_boost_round': 10}
Lowest MAE found:  4.370246171951294
time: 6min 9s (started: 2021-02-06 20:41:17 +00:00)


Best parameters found:  {'colsample_bytree': 0.8, 'early_stopping_rounds': 10, 'eta': 0.1, 'max_depth': 5, 'n_estimators': 100, 'num_boost_round': 10}

4.37

## Random Search

In [ ]:
from sklearn.model_selection import RandomizedSearchCV

# Create the parameter grid: here it corresponds to a range of values it can take from (distribution sort of)
gbm_param_grid = {
    'n_estimators': range(1, 1000),
    'max_depth': range(2, 12),
    'colsample_bytree': [0.1*i for i in range(1, 10)],
    'eta': [0.001*i for i in range(1, 10)] + [0.01*i for i in range(1, 10)] + [0.1*i for i in range(1, 10)],
}

# Instantiate the regressor: gbm
gbm = xgb.XGBRegressor()

# Perform random search: randomized_mse
randomized_mae = RandomizedSearchCV(param_distributions=gbm_param_grid, estimator=gbm, 
                                    scoring='neg_mean_absolute_error', n_iter=100, cv=2, 
                                   verbose=1, n_jobs=-1)

# Fit randomized_mse to the data
randomized_mae.fit(X_train, y_train)

# Print the best parameters and lowest RMSE
print("Best parameters found: ", randomized_mae.best_params_)
print("Lowest MAE found: ", np.abs(randomized_mae.best_score_))

In [ ]:
# Save to file
model = xgb.XGBRegressor(
    n_estimators= 12, 
    max_depth= 7, 
    eta= 0.06, 
    colsample_bytree=0.8,
)
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)
df_test['h_index_pred'].update(pd.Series(np.round_(y_pred, decimals=3)))
df_test.loc[:, ["authorID", "h_index_pred"]].to_csv(
    'predictions.csv', index=False
)